README.md
![Whisper](https://github.com/hyperoslo/Whisper/blob/master/Resources/whisper-cover.png)

![CircleCI](https://circleci.com/gh/hyperoslo/Whisper.png)
[![Version](https://img.shields.io/cocoapods/v/Whisper.svg?style=flat)](http://cocoadocs.org/docsets/Whisper)
[![Carthage](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![License](https://img.shields.io/cocoapods/l/Whisper.svg?style=flat)](http://cocoadocs.org/docsets/Whisper)
[![Platform](https://img.shields.io/cocoapods/p/Whisper.svg?style=flat)](http://cocoadocs.org/docsets/Whisper)
![Swift](https://img.shields.io/badge/%20in-swift%204.0-orange.svg)

## Description :leaves:

Break the silence of your UI, whispering, shouting or whistling at it. **Whisper** is a component that will make the task of display messages and in-app notifications simple. It has three different views inside.

#### Whispers

![Whisper](https://github.com/hyperoslo/Whisper/blob/master/Resources/permanent-whisper.png)

Display a short message at the bottom of the navigation bar, this can be anything, from a "Great Job!" to an error message. It can have images or even a loader.

#### Shouts

![In-App](https://github.com/hyperoslo/Whisper/blob/master/Resources/in-app-notification.png)

Let the users know that something happened inside the app with this beautiful customizable in app notification.

#### Whistles

![Whistle](https://github.com/hyperoslo/Whisper/blob/master/Resources/whistle-information.png)

This is the smallest of all, a beautiful discretion in your UI.

##### Bonus

All the sounds are fully customizable, from colors to fonts.

Shouts have an optional action that will be called if the user taps on it, and you'll even get a message when the Shout is gone. Finally, if you want to set how long the Shout should be displayed, you have a duration property.

In Whisper, there is no need to think about scroll view insets anymore, this will be handled automatically. As and added bonus, when transitioning from one view controller to another, the next controllers offset will be adjusted like you would expect. It just works!

## Usage

The usage of the component is so simple, you just create a message in the case of Whisper, an announcement in the case of a Shout or a Murmur in the case of a Whistle. Because there maybe conflict with `show` from `UIViewController`, you need to explicitly use `Whisper` namespace to call `show`

##### For a Whisper:

```swift
let message = Message(title: "Enter your message here.", backgroundColor: .red)

// Show and hide a message after delay
Whisper.show(whisper: message, to: navigationController, action: .show)

// Present a permanent message
Whisper.show(whisper: message, to: navigationController, action: .present)

// Hide a message
Whisper.hide(whisperFrom: navigationController)
```

##### For a Shout:

```swift
let announcement = Announcement(title: "Your title", subtitle: "Your subtitle", image: UIImage(named: "avatar"))
Whisper.show(shout: announcement, to: navigationController, completion: {
  print("The shout was silent.")
})
```

##### For a Whistle:

```swift
let murmur = Murmur(title: "This is a small whistle...")

// Show and hide a message after delay
Whisper.show(whistle: murmur, action: .show(0.5))

// Present a permanent status bar message
Whisper.show(whistle: murmur, action: .present)

// Hide a message
Whisper.hide(whistleAfter: 3)
```

If you want to use **Whisper** with Objective-C, you can find information about it [here](https://github.com/hyperoslo/Whisper/wiki/Using-Whisper-in-Objective-C).

## Installation

**Whisper** is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod 'Whisper'
```

**Whisper** is also available through [Carthage](https://github.com/Carthage/Carthage). To install just write into your Cartfile:

```ruby
github "hyperoslo/Whisper"
```

## Roadmap

In the future the idea is to keep improving and add some features:

- Improve the offset detection and animation.
- Add more UI related components into Whisper.
- More customization points and more sizes for each whisper.
- Custom actions inside Whispers and Shouts.
- We are open to new and awesome ideas, contribute if you like! :)

## Author

[Hyper](http://hyper.no) made this with ❤️

## Contribute

We would love for you to contribute to **Whisper**, check the [CONTRIBUTING](https://github.com/hyperoslo/Whisper/blob/master/CONTRIBUTING.md) file for more info.

## License

**Whisper** is available under the MIT license. See the [LICENSE](https://github.com/hyperoslo/Whisper/blob/master/LICENSE.md) file for more info.


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/hyperoslo/Whisper/tree/HEAD)

[Full Changelog](https://github.com/hyperoslo/Whisper/compare/3.0.0...HEAD)

**Merged pull requests:**

- fixed typo [\#84](https://github.com/hyperoslo/Whisper/pull/84) ([dawoodi](https://github.com/dawoodi))

## [3.0.0](https://github.com/hyperoslo/Whisper/tree/3.0.0) (2016-05-03)
[Full Changelog](https://github.com/hyperoslo/Whisper/compare/2.1...3.0.0)

**Merged pull requests:**

- Feature: permanent whistle [\#83](https://github.com/hyperoslo/Whisper/pull/83) ([vadymmarkov](https://github.com/vadymmarkov))
- Use assertFailure instead of fatalError [\#81](https://github.com/hyperoslo/Whisper/pull/81) ([onmyway133](https://github.com/onmyway133))
- Update to use \#selector [\#75](https://github.com/hyperoslo/Whisper/pull/75) ([onmyway133](https://github.com/onmyway133))

## [2.1](https://github.com/hyperoslo/Whisper/tree/2.1) (2016-03-18)
[Full Changelog](https://github.com/hyperoslo/Whisper/compare/2.0.2...2.1)

**Closed issues:**

- Button Actions in Shout / Announcement [\#67](https://github.com/hyperoslo/Whisper/issues/67)
- Feature: Ability to define timeout [\#60](https://github.com/hyperoslo/Whisper/issues/60)
- Build fails on Xcode 7.3 beta 2 [\#59](https://github.com/hyperoslo/Whisper/issues/59)
- Question: Customizing Shout [\#54](https://github.com/hyperoslo/Whisper/issues/54)
- Whistle's content is cropped [\#51](https://github.com/hyperoslo/Whisper/issues/51)

**Merged pull requests:**

- allow fonts to be configurable [\#69](https://github.com/hyperoslo/Whisper/pull/69) ([Shrugs](https://github.com/Shrugs))
- Error with hiding [\#64](https://github.com/hyperoslo/Whisper/pull/64) ([bevbomb](https://github.com/bevbomb))
- Fix view controller index typecast error [\#63](https://github.com/hyperoslo/Whisper/pull/63) ([andreyz](https://github.com/andreyz))
- Computes Murmur's height dynamically [\#53](https://github.com/hyperoslo/Whisper/pull/53) ([acadet](https://github.com/acadet))

## [2.0.2](https://github.com/hyperoslo/Whisper/tree/2.0.2) (2016-01-15)
[Full Changelog](https://github.com/hyperoslo/Whisper/compare/2.0.1...2.0.2)

**Merged pull requests:**

- Add missing validation of modifyInset in performControllerMove [\#49](https://github.com/hyperoslo/Whisper/pull/49) ([vadymmarkov](https://github.com/vadymmarkov))
- Add Config with modifyInset flag [\#48](https://github.com/hyperoslo/Whisper/pull/48) ([vadymmarkov](https://github.com/vadymmarkov))

## [2.0.1](https://github.com/hyperoslo/Whisper/tree/2.0.1) (2016-01-15)
[Full Changelog](https://github.com/hyperoslo/Whisper/compare/2.0.0...2.0.1)

**Merged pull requests:**

- Move controller up only if it's the same one that has been moved down [\#47](https://github.com/hyperoslo/Whisper/pull/47) ([vadymmarkov](https://github.com/vadymmarkov))

## [2.0.0](https://github.com/hyperoslo/Whisper/tree/2.0.0) (2016-01-15)
[Full Changelog](https://github.com/hyperoslo/Whisper/compare/1.1.0...2.0.0)

**Closed issues:**

- Feature: Ability to change text colour [\#36](https://github.com/hyperoslo/Whisper/issues/36)
- CONTRIBUTING.md references ImagePicker not Whisper documentation [\#35](https://github.com/hyperoslo/Whisper/issues/35)
- How to use this framework with objective-c? [\#28](https://github.com/hyperoslo/Whisper/issues/28)
- Whisper effect it's superview when it's called to show in viewDidLoad or viewDidAppear [\#22](https://github.com/hyperoslo/Whisper/issues/22)

**Merged pull requests:**

- Update podspec with new version [\#46](https://github.com/hyperoslo/Whisper/pull/46) ([zenangst](https://github.com/zenangst))
- Update code example for Whisper [\#45](https://github.com/hyperoslo/Whisper/pull/45) ([zenangst](https://github.com/zenangst))
- Feature text color for whispers [\#44](https://github.com/hyperoslo/Whisper/pull/44) ([zenangst](https://github.com/zenangst))
- Configure Travis to build the project [\#43](https://github.com/hyperoslo/Whisper/pull/43) ([zenangst](https://github.com/zenangst))
- Add Using Whisper in Objective-C to README [\#42](https://github.com/hyperoslo/Whisper/pull/42) ([zenangst](https://github.com/zenangst))
- Remove resource bundle from podspec [\#41](https://github.com/hyperoslo/Whisper/pull/41) ([vadymmarkov](https://github.com/vadymmarkov))
- Fixes Whistle [\#40](https://github.com/hyperoslo/Whisper/pull/40) ([RamonGilabert](https://github.com/RamonGilabert))
- Fix whistle messing with status bar appearance + fix indentation [\#39](https://github.com/hyperoslo/Whisper/pull/39) ([duemunk](https://github.com/duemunk))
- Actually use murmur duration for whistles [\#38](https://github.com/hyperoslo/Whisper/pull/38) ([duemunk](https://github.com/duemunk))

## [1.1.0](https://github.com/hyperoslo/Whisper/tree/1.1.0) (2016-01-05)
[Full Changelog](https://github.com/hyperoslo/Whisper/compare/1.0...1.1.0)

**Closed issues:**

- How to change Shout color [\#26](https://github.com/hyperoslo/Whisper/issues/26)
- Status Bar [\#25](https://github.com/hyperoslo/Whisper/issues/25)
- Application will be crashed if call Silent but never call Whisper before [\#23](https://github.com/hyperoslo/Whisper/issues/23)
- Whisper frame is not adjusted on orientation change [\#19](https://github.com/hyperoslo/Whisper/issues/19)
- Design awesome icon [\#2](https://github.com/hyperoslo/Whisper/issues/2)

**Merged pull requests:**

- Version 1.1 [\#34](https://github.com/hyperoslo/Whisper/pull/34) ([RamonGilabert](https://github.com/RamonGilabert))
- Broken layout [\#33](https://github.com/hyperoslo/Whisper/pull/33) ([RamonGilabert](https://github.com/RamonGilabert))
- Shout colors [\#32](https://github.com/hyperoslo/Whisper/pull/32) ([RamonGilabert](https://github.com/RamonGilabert))
- ShoutFactory: hiding setters from the public API [\#31](https://github.com/hyperoslo/Whisper/pull/31) ([NachoSoto](https://github.com/NachoSoto))

Source/ShoutFactory.swift
import UIKit

let shoutView = ShoutView()

open class ShoutView: UIView {

  public struct Dimensions {
    public static let indicatorHeight: CGFloat = 6
    public static let indicatorWidth: CGFloat = 50
    public static let imageSize: CGFloat = 48
    public static let imageOffset: CGFloat = 18
    public static var textOffset: CGFloat = 75
    public static var touchOffset: CGFloat = 40
  }

  open fileprivate(set) lazy var backgroundView: UIView = {
    let view = UIView()
    view.backgroundColor = ColorList.Shout.background
    view.alpha = 0.98
    view.clipsToBounds = true

    return view
    }()

  open fileprivate(set) lazy var indicatorView: UIView = {
    let view = UIView()
    view.backgroundColor = ColorList.Shout.dragIndicator
    view.layer.cornerRadius = Dimensions.indicatorHeight / 2
    view.isUserInteractionEnabled = true

    return view
    }()

  open fileprivate(set) lazy var imageView: UIImageView = {
    let imageView = UIImageView()
    imageView.layer.cornerRadius = Dimensions.imageSize / 2
    imageView.clipsToBounds = true
    imageView.contentMode = .scaleAspectFill

    return imageView
    }()

  open fileprivate(set) lazy var titleLabel: UILabel = {
    let label = UILabel()
    label.font = FontList.Shout.title
    label.textColor = ColorList.Shout.title
    label.numberOfLines = 2

    return label
    }()

  open fileprivate(set) lazy var subtitleLabel: UILabel = {
    let label = UILabel()
    label.font = FontList.Shout.subtitle
    label.textColor = ColorList.Shout.subtitle
    label.numberOfLines = 2

    return label
    }()

  open fileprivate(set) lazy var tapGestureRecognizer: UITapGestureRecognizer = { [unowned self] in
    let gesture = UITapGestureRecognizer()
    gesture.addTarget(self, action: #selector(ShoutView.handleTapGestureRecognizer))

    return gesture
    }()

  open fileprivate(set) lazy var panGestureRecognizer: UIPanGestureRecognizer = { [unowned self] in
    let gesture = UIPanGestureRecognizer()
    gesture.addTarget(self, action: #selector(ShoutView.handlePanGestureRecognizer))

    return gesture
    }()

  open fileprivate(set) var announcement: Announcement?
  open fileprivate(set) var displayTimer = Timer()
  open fileprivate(set) var panGestureActive = false
  open fileprivate(set) var shouldSilent = false
  open fileprivate(set) var completion: (() -> ())?

  private var subtitleLabelOriginalHeight: CGFloat = 0
  private var internalHeight: CGFloat = 0

  // MARK: - Initializers

  public override init(frame: CGRect) {
    super.init(frame: frame)

    addSubview(backgroundView)
    [imageView, titleLabel, subtitleLabel, indicatorView].forEach {
      $0.autoresizingMask = []
      backgroundView.addSubview($0)
    }

    clipsToBounds = false
    isUserInteractionEnabled = true
    layer.shadowColor = UIColor.black.cgColor
    layer.shadowOffset = CGSize(width: 0, height: 0.5)
    layer.shadowOpacity = 0.1
    layer.shadowRadius = 0.5

    backgroundView.addGestureRecognizer(tapGestureRecognizer)
    addGestureRecognizer(panGestureRecognizer)

    NotificationCenter.default.addObserver(self, selector: #selector(ShoutView.orientationDidChange), name: NSNotification.Name.UIDeviceOrientationDidChange, object: nil)
  }

  public required init?(coder aDecoder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  deinit {
    NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIDeviceOrientationDidChange, object: nil)
  }

  // MARK: - Configuration

  open func craft(_ announcement: Announcement, to: UIViewController, completion: (() -> ())?) {
    panGestureActive = false
    shouldSilent = false
    configureView(announcement)
    shout(to: to)

    self.completion = completion
  }

  open func configureView(_ announcement: Announcement) {
    self.announcement = announcement
    imageView.image = announcement.image
    titleLabel.text = announcement.title
    subtitleLabel.text = announcement.subtitle

    displayTimer.invalidate()
    displayTimer = Timer.scheduledTimer(timeInterval: announcement.duration,
      target: self, selector: #selector(ShoutView.displayTimerDidFire), userInfo: nil, repeats: false)

    setupFrames()
  }

  open func shout(to controller: UIViewController) {
    controller.view.addSubview(self)

    frame.size.height = 0
    UIView.animate(withDuration: 0.35, animations: {
      self.frame.size.height = self.internalHeight + Dimensions.touchOffset
    })
  }

  // MARK: - Setup

  public func setupFrames() {
    internalHeight = (UIApplication.shared.isStatusBarHidden ? 55 : 65)

    let totalWidth = UIScreen.main.bounds.width
    let offset: CGFloat = UIApplication.shared.isStatusBarHidden ? 2.5 : 5
    let textOffsetX: CGFloat = imageView.image != nil ? Dimensions.textOffset : 18
    let imageSize: CGFloat = imageView.image != nil ? Dimensions.imageSize : 0

    [titleLabel, subtitleLabel].forEach {
        $0.frame.size.width = totalWidth - imageSize - (Dimensions.imageOffset * 2)
        $0.sizeToFit()
    }

    internalHeight += subtitleLabel.frame.height

    imageView.frame = CGRect(x: Dimensions.imageOffset, y: (internalHeight - imageSize) / 2 + offset,
      width: imageSize, height: imageSize)

    let textOffsetY = imageView.image != nil ? imageView.frame.origin.x + 3 : textOffsetX + 5

    titleLabel.frame.origin = CGPoint(x: textOffsetX, y: textOffsetY)
    subtitleLabel.frame.origin = CGPoint(x: textOffsetX, y: titleLabel.frame.maxY + 2.5)

    if subtitleLabel.text?.isEmpty ?? true {
      titleLabel.center.y = imageView.center.y - 2.5
    }

    frame = CGRect(x: 0, y: safeYCoordinate,
                   width: totalWidth, height: internalHeight + Dimensions.touchOffset)
  }

  // MARK: - Frame

  open override var frame: CGRect {
    didSet {
      backgroundView.frame = CGRect(x: 0, y: safeYCoordinate,
                                    width: frame.size.width,
                                    height: frame.size.height - Dimensions.touchOffset)

      indicatorView.frame = CGRect(x: (backgroundView.frame.size.width - Dimensions.indicatorWidth) / 2,
                                   y: backgroundView.frame.height - Dimensions.indicatorHeight - 5,
                                   width: Dimensions.indicatorWidth,
                                   height: Dimensions.indicatorHeight)
    }
  }

  // MARK: - Actions

  open func silent() {
    UIView.animate(withDuration: 0.35, animations: {
      self.frame.size.height = 0
      }, completion: { finished in
        self.completion?()
        self.displayTimer.invalidate()
        self.removeFromSuperview()
    })
  }

  // MARK: - Timer methods

    @objc open func displayTimerDidFire() {
    shouldSilent = true

    if panGestureActive { return }
    silent()
  }

  // MARK: - Gesture methods

  @objc fileprivate func handleTapGestureRecognizer() {
    guard let announcement = announcement else { return }
    announcement.action?()
    silent()
  }
  
  @objc private func handlePanGestureRecognizer() {
    let translation = panGestureRecognizer.translation(in: self)

    if panGestureRecognizer.state == .began {
      subtitleLabelOriginalHeight = subtitleLabel.bounds.size.height
      subtitleLabel.numberOfLines = 0
      subtitleLabel.sizeToFit()
    } else if panGestureRecognizer.state == .changed {
      panGestureActive = true
      
      let maxTranslation = subtitleLabel.bounds.size.height - subtitleLabelOriginalHeight
      
      if translation.y >= maxTranslation {
        frame.size.height = internalHeight + maxTranslation
          + (translation.y - maxTranslation) / 25 + Dimensions.touchOffset
      } else {
        frame.size.height = internalHeight + translation.y + Dimensions.touchOffset
      }
    } else {
      panGestureActive = false
      let height = translation.y < -5 || shouldSilent ? 0 : internalHeight

      subtitleLabel.numberOfLines = 2
      subtitleLabel.sizeToFit()
      
      UIView.animate(withDuration: 0.2, animations: {
        self.frame.size.height = height + Dimensions.touchOffset
      }, completion: { _ in
          if translation.y < -5 {
            self.completion?()
            self.removeFromSuperview()
        }
      })
    }
  }


  // MARK: - Handling screen orientation

    @objc func orientationDidChange() {
    setupFrames()
  }
}
